Resizable UI

Endaris opened the pull request:
Work-in-Progress branch for overhauling our UI (again)

# Goal
Move away from the 16:9 static global canvas and make it so that scenes and UI can be designed in a way that looks okay on mobile without compromising much on design for desktop.

 # Changes

## Font handling

Font sizes used to be handled in numeric sizes and the way the size was set was inconsistent across the repository.

The new font handling has a range of string keys that refer to specific base sizes so that different text can be sized relatively to each other using those keys but without requiring explicit numbers.
Instead the GraphicsUtil font functions now access an extra parameter `fontSizeDelta` that is added to the standard font size for each key before fetching the font.
This has the advantage that
1. Theme/Language fonts using off-standard sizes no longer suffer from divergent relative sizing when bigger/smaller fonts than the standard were sometimes addressed in absolute pt sizes.
2. Font sizes can easily be increased on mobile without bringing the handling into the Label element.

## FlexBox Layout

This is mostly described within the PR in the updated `Understanding the Codebase.md` doc so read there instead.

## New Battle Room UI

This is under construction.
The idea is to move to a vertical layout that makes use of scrollable and wrappable elements to provide much more screen real estate to implement more sophisticated UI options going forward, in many cases UI options that cannot be properly fit into the current battle room UI.

# TODO

- [ ] Repair resolution-aware use of dpi scale for assets
- [ ] Maybe rethink how player cursor focus is done on UIElements that are navigable by more than one player
- [ ] Extract directly window related tasks into its own component

Endaris committed:
add layouting logic and adapt it for UiElement

Endaris committed:
make specialized scrollcontainer to replace Menu

Endaris committed:
some layouting adjustments for options menu

Endaris committed:
fully replace Menu with new ScrollContainer-based VerticalMenu

Endaris committed:
rework font handling to work with discrete key sizes

Endaris committed:
add adaptive layout and fix some more font stuff

Endaris committed:
fix a crash, fix graphics menu adjustment, fix slider alignment

Endaris committed:
refix slider positioning
fix an issue where the top level children would not grow in the direction of the layout orientation

Endaris committed:
vaguely resolve most crashes for a slightly broken layout (lobby/vsSelf)

Endaris committed:
add a static layout that blocks flex propagation

Endaris committed:
adjust layouting to use a UIElement specific baseWidth/Height

Endaris committed:
add shrinking to horizontalFlexLayout

Endaris committed:
abandon fixed canvas resolution, embrace dynamic window size

Endaris committed:
add a very basic standard cursor for whatever needs to kickstart a nested focuse navigation

Endaris committed:
initial cursor display for designhelper

Endaris committed:
create a container that unifies child size allows grid size navigation and automatically wraps for the given width

Endaris committed:
Merge branch 'beta' into resizableUi

Endaris committed:
fix most things about wrapping containers with children

Endaris committed:
fix scene.uiRoot maximum dimensions
fix childGaps not being subtracted when calculating remaining width to distribute

Endaris committed:
change alignment options to be parent based
fix annotations for ui/init.lua

Endaris committed:
fix alignment for unisizedcontainer children and default to minWidth if the window is squeezed too small instead of the preferred width

Endaris committed:
fix a navigation crash when unisizedcontainer used vAlign different from top

Endaris committed:
update Understanding the Codebase doc with info about the new layout mechanism and new GameMode related enums

Endaris committed:
extract CharacterButton from CharacterSelect into its own UIElement

Endaris committed:
update doc how localization is implemented in the Label UIElement

JamesVanBoxtel commented on client/src/Game.lua:
```diff
@@ -354,13 +350,22 @@ function Game:updateMouseVisibility(dt)
 end
 
 function Game:handleResize(newWidth, newHeight)
+  local activeScene = GAME.navigationStack:getActiveScene()
```
There should probably be a root element that all scenes are parented from, that way we can draw overlays on it etc and it is always the root spot to update scale / height / width on.

Endaris replied:
Yep!
This is still WIP and I did this quick and dirty to get a layout resize for window resizes so I could test the layout at all. I intend to come back to this.

JamesVanBoxtel commented on client/src/Game.lua:
```diff
@@ -653,14 +645,19 @@ function Game:setLanguage(lang_code)
   config.language_code = Localization.codes[Localization.lang_index]
 
+  local baseOffset = 0
+  if system.isMobileOS() or DEBUG_ENABLED then
```
Sprinkling these DEBUG_ENABLED make it hard to test because running in debug mode you don't know what behavior is changing. Could we make a generic "test mobile" debug flag that we could set from the debug menu?

Endaris replied:
Sure, done!

JamesVanBoxtel reviewed (commented):
What is your plan for brining back the black letter boxing border on desktop?

Maybe we add the navigation stack to a sub view that is centered and have the black background on the root view? Then eventually all these calls to canvas width can instead get the size of the appropriate parent element or use the new fill alignment layout stuff.

About half way done looking through code, still need to test.

Endaris committed:
add a new command line flag to simulate mobile OS instead of doing it by default in debug mode

Endaris commented:
> What is your plan for bringing back the black letter boxing border on desktop?
> 
> Maybe we add the navigation stack to a sub view that is centered and have the black background on the root view? Then eventually all these calls to canvas width can instead get the size of the appropriate parent element or use the new fill alignment layout stuff.

Letterboxing specifically exists to fill remaining space when the window size doesn't match the desired aspect ratio but with the new layout strategy it could technically just adapt to that size as we no longer have a specific desired aspect ratio. 
`Game:draw` already clears the background to black so with some luck, some kind of "letterboxing" will naturally emerge where the background image does not reach.
I believe it will be easier to evaluate this once we have working prototypes for most scenes so it's easier to see how our scenes pan out and where/how letterboxing would be most desirable.

Endaris committed:
no layouting

Endaris committed:
start working on an integrated cursor system and fix some luaLS intellisense stuff

Endaris committed:
widely use import with the luaLS plugin so it properly resolves relative requires

Endaris committed:
change receiveInputs to be cursor driven to achieve a UI system in which relevant UI elements always know whether they are being hovered by a keyboard/controller cursor
cursors can now also traverse focus within a scene in a non-linear (aka a button can pass it to an entirely different UiElement) and reversible way

Endaris committed:
fix getScreenPos() for children living in ScrollContainers not yielding the actual on-screen position

Endaris committed:
fix touch detection in ScrollContainers and mostly unbrick DesignHelper

Endaris committed:
some cleanup

Endaris committed:
more cleanup

Endaris committed:
port client parts of consts.lua to its own file

Endaris committed:
wrap characterSelect buttons in scrollcontainer and fix a bug that caused children to get repositioned twice, causing failure for wrappables inside of scrollContainers

Endaris committed:
fix characterbutton offsets

Endaris committed:
fix CharacterButton display

Endaris committed:
fix cursor navigation on UniSizedContainer

Endaris committed:
only show super select stage while a character is hovered (cause it can look bad as default)

Endaris committed:
add PanelSetButton

Endaris committed:
fix some things regarding resizing with overflowing height

JamesVanBoxtel commented on client/src/ui/UniSizedContainer.lua:
```diff
@@ -0,0 +1,231 @@
+---@field selectedColumn integer
+---@field rows UiElement[][]
+---@overload fun(options: UniSizedContainerOptions): UniSizedContainer
+local UniSizedContainer = class(
```
Could you add documentation on what this element is intended for?

It has rows and next? Is this a grid or something?

JamesVanBoxtel commented on client/src/ui/UIElement.lua:
```diff
@@ -78,34 +112,47 @@ local UIElement = class(
+        self.layout.resize(self)
+      --end
+      if oWidth ~= self.width or oHeight ~= self.height then
+        GraphicsUtil.updateMode(self.width, self.height, {})
```
I don't get why the root element would drive the window size? Shouldn't the window size be driven by the OS / device / window / game options?

Endaris replied:
The root element has its maximum width constrained by the desktop dimensions and the stated minimum is controlled by what the user / OS does with the window.
What this condition + updateMode does is enforce a minimum window width, like how I cannot resize my browser to a 1x1 window but instead it stops here:
![grafik](https://github.com/user-attachments/assets/0c0ff831-1dbc-4465-9cef-ab3e364e60f4)

There are two instances of this check + updateMode, one when the user actively tries to shrink the window and another when children are added because the minimum dimensions of the window could have changed with the new child in which case it has to grow.

These are certainly constraints that should maybe not be in UIElement and rather part of an element that explicitly deals with window concerns like the letterboxing you mentioned, but I'm happy to leave it here until I have the rest of UI and cursor navigation figured out.
Added a todo in the opening post of the PR!

/edit:
Actually there was a bug here, `resize` has to be called with the current window size instead of empty args so it sizes to its current width rather than fits to the window.

JamesVanBoxtel commented on client/src/ui/UIElement.lua:
```diff
@@ -114,6 +161,7 @@ function UIElement:detach()
       if child.id == self.id then
         table.remove(self.parent.children, i)
         self:onDetach()
+        self.parent:onChildrenChanged()
```
Might want to make a setParent method that calls on children changed so the code is shared for future changes.

Endaris replied:
It's only 2 calls and I don't see it ever being more so kind of YAGNI, if we ever upheave our UI again, this is going to be our smallest problem.
Actually though, outside elements should never have a reason to call `onChildrenChanged` or alternatively a `setParent` directly so I'll actually turn that function into a local so it does not clutter intellisense.

JamesVanBoxtel commented on client/src/ui/UIElement.lua:
```diff
@@ -159,14 +205,22 @@ end
 -- UiElements containing children draw the children-independent part in this function
 -- implementation is optional so layout elements don't have to
 function UIElement:drawSelf()
+  love.graphics.setColor(self.backgroundColor)
```
Note we still have this weird design pattern where drawSelf needs to take into account x, y

I still think the transform should be applied first so you can draw in local coordinates instead of at x y, but we can do that later.

Endaris replied:
Agreed, I think previously that was a bit odd due to alignment being child-driven and x/y being set to values that depended on the alignment (e.g. right-aligned elements often had negative x) but there's certainly no reason to not change it now.
I'll hold it for a bit because it will involve adjusting drawSelf for most elements.
/edit: done actually

JamesVanBoxtel commented on client/src/ui/Layouts/VerticalFlexLayout.lua:
```diff
@@ -0,0 +1,170 @@
+---@return number # the minimum width of the element as dictated by its children
+function VerticalFlexLayout.getMinWidth(uiElement)
+  local w = uiElement.padding * 2
+  local maxWidth = 0
```
variable name should be minWidth?

Endaris replied:
The minimum width of a vertically laid out flex box is the width of its widest element plus the padding.
So `maxWidth` refers to the maximum width among the children specifically.
`maxChildWidth` is probably more intuitive so I'll rename this.

JamesVanBoxtel commented on client/src/ui/ImageContainer.lua:
```diff
@@ -36,10 +38,10 @@ function ImageContainer:setImage(image, width, height, scale)
   self.scale = math.min(self.width / self.imageWidth, self.height / self.imageHeight)
-  self.width = self.imageWidth * self.scale
-  self.height = self.imageHeight * self.scale
+  -- self.width = self.imageWidth * self.scale
```
Delete?

JamesVanBoxtel commented on client/src/scenes/TrainingMenu.lua:
```diff
@@ -84,39 +82,52 @@ function TrainingMenu:load(sceneParams)
+    local go = ui.MenuItem.createButtonMenuItem("go_", nil, nil, function() self:goToCharacterSelect(garbagePatternStepper.value, widthSlider.value, heightSlider.value) end)
+    local back = ui.MenuItem.createButtonMenuItem("back", nil, nil, exitMenu)
+
+  self.menu = ui.VerticalMenu({
```
I still think we should have a constructor here that does "the right thing" for main menu menus so we can adjust in one spot.

Endaris replied:
I guess, to be fair this is really placeholderish.
Looking at my current idea I want to support in-room game mode adjustments so this particular one will most likely be embedded later but I get what you mean in general.
I'll keep that in mind for when I try to establish the general outline for the "split screen" idea.

Endaris committed:
4-slice the cursor image and draw the cursor properly

Endaris committed:
fix vertical resizing for preferred vs min height
apply some of Jam's feedback in comments / var renames
4-slice the image cursor and draw it with maintained aspect ratio

Endaris committed:
make onChildrenChanged local

Endaris committed:
change UIElement:drawSelf and all its override to assume 0, 0 offset and move the translate in UIElement:draw

Endaris committed:
readjust MenuItem
explicitly use cursors for Options/SoundTest

Endaris committed:
update doc

Endaris committed:
divide Button clicks into action and onAction
start on stage selection

Endaris committed:
fix some stuff about images and stagebuttons
